### PR TITLE
ARROW-6360: [R] Update support for compression

### PR DESCRIPTION
**Pull request by romainfrancois**
codec = NULL means choose the right default for the current platform. Default is set for `SNAPPY` on windows and `GZIP` otherwise, instead of universally using `GZIP` as the default and abort on windows.

### CODE REVIEW

**Comment by pitrou** [comment]
I don't want to opiniate too much on this (not a R developer), but I'm curious why the choice of snappy (which AFAIK is little used nowadays, as there are better options). Also, why would gzip fail on Windows? Gzip is so universal that hopefully it should be supported on Windows as well.

**Comment by nealrichardson** [comment]
It all depends on what the C++ library is built with on Windows. According to https://github.com/apache/arrow/blob/master/ci/PKGBUILD#L88, gzip should be there. 

Snappy support was added specifically because it was (reported to be) common with Parquet files.

Speaking of Parquet, as the ticket said it would be nice to support writing to parquet with compression (cf. https://github.com/apache/arrow/pull/5192, which further added support to the C++ library for specifying a compression level).

**Comment by pitrou** [comment]
Right, so personally I would make gzip the default everywhere, if you do need a default.

**Comment by nealrichardson** [comment]
Agreed.

**Comment by wesm** [comment]
Snappy should be the default compression when writing Parquet files. Make sure it isn't gzip or benchmarks will suffer (almost every library out there defaults to Snappy AFAIK)

But anywhere else where there is a need for a compression gzip seems like the reasonable default

**Review by nealrichardson — COMMENTED** [review]
I rebased and pushed a few changes. I guess this R code is covered by test-compression.R, and the real business is covered in the C++ library elsewhere. 

If it makes sense, you could separate the parquet encryption support to a different ticket and merge this.

File: r/data-raw/codegen.R [hunk 1]
```diff
@@ -53,7 +54,7 @@ decorations <- cpp_decorations() %>%
   # more concisely
   # rap(            ~ decor:::parse_cpp_function(context))
   mutate(functions = map(context, decor:::parse_cpp_function)) %>%
-  { bind_cols(., bind_rows(pull(., functions))) } %>%
```
[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/data-raw/codegen.R]: Nevermind, I see we already `Suggest` `vctrs`.

**Comment by nealrichardson** [comment]
+1. I created https://issues.apache.org/jira/browse/ARROW-6532 and https://issues.apache.org/jira/browse/ARROW-6533 for followups.

**Pull request closed without merging**